### DISCUSSION TITLE
ISR: Include Hostname in getStaticProps

### DISCUSSION THREAD

**Discussion by Afulton11** [vercel/next.js#20116] [Ideas]
**ISR** - Incremental Static (Re)generation, See [Next.JS 9.5](https://nextjs.org/blog/next-9-5#stable-incremental-static-regeneration)

## Feature Request/Goal

- Provide the ability to cache pages using a combination of the pathname AND hostname for pages rendered with ISR.
  - Access the hostname within `getStaticProps({...})`

## Background

With Vercel's ability to [wildcard subdomains](https://vercel.com/blog/wildcard-domains) for Next.JS applications, we can server-side and client-side render pages depending on the subdomain. Yet, we can't render pages using ISR depending on the subdomain. For my particular use case, I want to serve different data depending on the subdomain a user is visiting.

There have already been a few discussions created with people asking about this issue:

- https://github.com/vercel/next.js/discussions/12848
- https://github.com/vercel/next.js/discussions/18142
- https://github.com/vercel/next.js/discussions/20051

## The Problem

With wildcard domains, the subdomain changes _all_ the time. Depending on the application, One may want their app to show different data depending on the unknown (at build-time) subdomain.

In Next.Js, the current cache for ISR stores the rendered page by the URL's pathname (e.g. for a full URL of `example.hello.com/user/afulton11`, the pathname would be `/user/afulton11`). With the current implementation of `getStaticProps(...)`, using this as the cache-key works well.

However, we run into issues when the parts of the URL before the path change. Even using a workaround to fetch the subdomain within `getStaticProps(..)` will not work because the cache will serve the last statically-generated page no matter the subdomain (except for domain routed locales).

### ISR Data Fetching

```js
export async function getStaticProps({ hostname, ...rest }) {
  const [subdomain, domain] = hostname?.split('.')

  const hostSpecificData = await fetch(`https://.../some-api?name=${subdomain}`)

  return {
    props: {
      hostSpecificData,
    },
  }
}
```

I am also open to using some configuration option to enable `host-specific ISR`.

## Potential Implementation

1. The [host](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Host) header contains the hostname of the current request.

   - e.g. `example.subdomain.cool.com`
   - Next.JS is already using the `host` header for `locale` support, see [RFC: I18n Routing](https://github.com/vercel/next.js/discussions/17078) and [next-server.ts#L328](https://github.com/vercel/next.js/blob/c792317295c076e0b449ada10571b0716e395111/packages/next/next-server/server/next-server.ts#L328)

2. Use the same method as locale ([next-server.ts#L328](https://github.com/vercel/next.js/blob/c792317295c076e0b449ada10571b0716e395111/packages/next/next-server/server/next-server.ts#L328)) to parse `hostname` from the request.
3. Add `hostname` to the `ssgCacheKey` in `next-server.ts`
4. Add `hostname` to `renderOptsPartial` in `render.tsx`
5. Add `hostname` to `getStaticProps({...})` in [render.tsx#L614](https://github.com/vercel/next.js/blob/c792317295c076e0b449ada10571b0716e395111/packages/next/next-server/server/render.tsx#L614)

### Additional Context

There are many different ways to solve this problem in Next.JS; this is one of the more straightforward solutions I found.

#### Some Questions
* Is this something that would be useful to have in Next.JS? 
* Are there any downsides with adding the hostname to the ISR cache key & `getStaticProps` parameters?
* Is there a better way to access host data during ISR?
* Do you have a different idea on how to tackle this problem in NextJS? 
* Any comments or criticisms? 

Let me know what you think!

**Comment by raphaelbadia** [comment]
I'd love to see this implemented !

**Comment by nicosh** [comment]
@timneutkens Any plan on release this feature?

**Comment by nicosh** [comment]
@Afulton11 Another solution could be this : 
Allow to have an option on `next.config.js` called something like `isgPathPrefix` default false if true default pages are not retrivied from 
 `./next/server/pages` but from `./next/server/pages/[pathprefix]` .
 Then in your page you can specify path prefix in this way : 
 
 With ISG : 
 ```
 // if isgPathPrefix is enabled we pass a function for retrieve hostname and use it as path prefix
 export async function getStaticPaths() {
  return { paths: [], fallback: true, getPathPrefix = req => req.hostname };
}

// then in getStaticProps context params contain not obly path but also pathprefix
export const getStaticProps = async ({params}) => {
  let data = await loadPage(params.pathPrefix,params.path);
  return {
    props: {
      data
    },
    revalidate: 1, // In seconds
  }
}
 ```
 
 Without ISG we can just pass pathprefix to `getStaticPaths`
 
 ```
 const data = [ { params: { id: '1',pathPrefix : "website1.com" }]
 export async function getStaticPaths() {
  return {
    paths:  data,
    fallback: true or false 
  };
}
 ```
 
Not sure if this make sense or is doable, but still that retrieve hostname when regenerating pages would be a nice to have feature.

**Comment by styxlab** [comment]
Would like to see this idea materialize!

**Comment by tmm** [comment]
Linking related post on Reddit https://www.reddit.com/r/nextjs/comments/ms56mr/isr_with_custom_subdomains/

**Reply by SleepWalker** [reply]
And one more discussion with updates regarding next@10.2 and new redirect features:

https://github.com/vercel/next.js/discussions/12848#discussioncomment-680225

**Comment by neenhouse** [comment]
+1 framework support for multi-host ISR would be very nice without having to do maintain work-around host rewrite configurations or custom middleware.

@Afulton11 which approach have you settled on?

**Comment by ciandt-crodrigues** [comment]
Man, feels bad this is going on for such a long time.
it seems like a really simple thing to add and would solve so many problems...
